packages/block-library/src/gallery/gallery.native.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import { View } from 'react-native';
+
+/**
+ * Internal dependencies
+ */
+import GalleryImage from './gallery-image';
+import { defaultColumnsNumber } from './shared';
+import styles from './gallery-styles';
+import Tiles from './tiles';
+
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+
+const TILE_SPACING = 15;
+
+export const Gallery = ( props ) => {
+	const {
+		selectedImage,
+		mediaPlaceholder,
+		onMoveBackward,
+		onMoveForward,
+		onRemoveImage,
+		onSelectImage,
+		onSetImageAttributes,
+		//	onFocusGalleryCaption,
+		attributes,
+		isSelected,
+		//	setAttributes,
+		isMobile,
+		isNarrow,
+		onFocus,
+	} = props;
+
+	const {
+		columns = defaultColumnsNumber( attributes ),
+		imageCrop,
+		images,
+	} = attributes;
+
+	let displayedColumns;
+
+	// limit displayed columns based on viewport width
+	if ( isMobile ) {
+		displayedColumns = Math.min( columns, 2 );
+	} else if ( isNarrow ) {
+		displayedColumns = Math.min( columns, 4 );
+	} else {
+		displayedColumns = columns;
```

Comment:
I gave it a try with iPad Pro 11 inch simulator, unfortunately our readable content width is still too narrow for 8 columns. Our readable content max width is 580.(see .../src/mobile/readable-content-view/style.native.scss) I don't know how isNarrow is calculated but that calculation should be based on readable content width and not screen width.


<img width="892" alt="Screen Shot 2019-11-25 at 17 07 07" src="https://user-images.githubusercontent.com/5032900/69550878-daed5580-0fac-11ea-98de-6264bdefebe5.png">

Reply:
Thanks for pointing this out. I've pushed changes to set a `MAX_DISPLAYED_COLUMNS = 4` to limit the number of columns displayed for native. The displayed columns will still be limited to 2 for narrower screens.

Reply:
I am fine with that fix currently 👍 In the long term we might need a more dynamic decision mechanism which can adjust itself accordingly due to the width of readable content. But, currently we have a constant limit there. So I don't see enough value to introduce such a dynamic mechanism today.